#This script is designed to run with Understand - CodeCheck
# Written by Kyle Jackson

# Boiler Plate code
use base ("Understand::Codecheck");
use strict;

# This is the error that will be returned when a violation is found
use constant ERR1 => 'Class type exception not caught by reference in file %1';

# This registers ERR1 with Understand
sub register_tr_text() {
  my $check = shift;
  $check->add_tr_text(ERR1);
}

# This is the short error the Understand will use
sub name { return "A class type exception shall always be caught by reference";}

# This is the full description of the rule being tested
sub description { return "A class type exception shall always be caught by reference.";}

# This section returns the rational and examples to be given to the user through Understand
sub detailed_description { return <<"END_DESC"
<p><b>Rationale</b><br>
If a class type exception object is caught by value, slicing occurs. That is, if the exception object is of
a derived class and is caught as the base, only the base class’s functions (including virtual functions)
can be called. Also, any additional member data in the derived class cannot be accessed.</p>
<br>
<p>
If the exception is caught by reference, slicing does not occur.</p>
<br>
<b>Example</b><pre style="margin-top:0;padding-top:0;">
  // base class for exceptions
  class ExpBase
  {
  public:
    virtual const char_t *who ( )
    {
      return "base";
    };
  };
  class ExpD1: public ExpBase
  {
  public:
    virtual const char_t *who ( )
    {
      return "type 1 exception";
    };
  };
  class ExpD2: public ExpBase
  {
  public:
    virtual const char_t *who ( )
    {
      return "type 2 exception";
    };
  };

  int main() {
    try
    {
      // ...
      throw ExpD1 ( );
      // ...
      throw ExpBase ( );
    }
    catch ( ExpBase &b ) // Compliant – exceptions caught by reference
    {
      // ...
      b.who(); // "base", "type 1 exception" or "type 2 exception"
               // depending upon the type of the thrown object
    }
    // Using the definitions above ...
    catch ( ExpBase b ) // Non-compliant - derived type objects will be
                        // caught as the base type
    {
      b.who(); // Will always be "base"
      throw b; // The exception re-thrown is of the base class,
               // not the original exception type
    }
  }
</pre>
END_DESC
}

# This section defines the programming language this script is testing
sub test_language {
  my $language = shift;
  return $language =~ /C\+\+/; #Handles C and C++
}

# This determines if this script is per file
sub test_entity { return 1;}

# This determines if this script is per project or per file
sub test_global { return 0;}

# Any defined options go here
sub define_options{}

# Use this function for throwing errors
sub throwError {
  my $check = shift;
  my $ent = shift;
  my $file = shift;
  my $line = shift;
  my $column = shift;
  my $error = shift;
  my $name = shift;
  $check->violation($ent,$file,$line,$column,$error,$name);
}

# This is the actual test being performed on the user's code
sub check {
  # Pull the name of the check and the name of the file from the incoming script request
  my $check = shift; 
  my $file = shift;
  return unless $file->kind->check('c file');
  
  # Create a list of all classes that are defined
  my @functionsDefined = $file->filerefs("define");
  return unless @functionsDefined;
  
  # Loop through each function looking for the catch command
  foreach my $function(@functionsDefined) {
    # Check to see if the keyword catch exists in the text
	next unless $function->ent->contents =~ /catch/;
	
	# We have found a function with a catch block; Initialize and check the lexer
	my $lexer = $file->lexer(1,0,0);
    return unless $lexer;
	
	# Set the first lexeme and counters need to check entire function
	my $counter = 1;
	my $first = 0;
	my $block = 0;
    my $lexeme = $lexer->lexeme($function->line, $function->column);
	
	#While we are still inside the current function
	while ($lexeme && $lexeme->line_begin <= $function->ent->ref("end")->line){
	  $block = 1 if $lexeme->text eq "catch";
	  next unless $block == 1;
	  $counter++ if $first == 1 and $lexeme->text eq "{";
	  $counter-- if $first == 1 and $lexeme->text eq "}";
	  $first = 1 if $first == 0 and $lexeme->text eq "{";
	  next unless $lexeme->ent;
	  if ($lexeme->ent->kindname =~ /Class/ && !($lexeme->next->next->text eq "&") && !($lexeme->next->next->next->next->text eq "&")) {
	    throwError($check, $function->ent,$file,$lexeme->line_begin,$lexeme->column_begin,ERR1,$function->file->name);
	  }
	  if ($counter == 0) {
	    $block = 0;
		$counter = 1;
	  }
	}
	continue {
      $lexeme = $lexeme->next;
    }
  }
}

sub checkLexeme {
  my $lexeme ->shift;
}




















